include/openthread/random.h

```diff
@@ -0,0 +1,163 @@
+/*
+ *  Copyright (c) 2019, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * @brief
+ *  This file defines the OpenThread random number generator API.
+ */
+
+#ifndef OPENTHREAD_RANDOM_H_
+#define OPENTHREAD_RANDOM_H_
+
+#include "openthread-core-config.h"
+
+#include "utils/wrap_stdint.h"
+
+#include <mbedtls/ctr_drbg.h>
+#include <openthread/error.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @addtogroup api-random
+ *
+ * @brief
+ *   This module includes functions that generates random numbers.
+ *
+ * @{
+ *
+ */
+
+/**
+ * This function generates and returns a random `uint32_t` value.
+ *
+ * @returns    A random `uint32_t` value.
+ *
+ */
+uint32_t otRandomSoftwareGetUint32(void);
```

Comment:
I understand you are trying to differentiate from a cryptographic PRNG. However, I think it's strange to use `Software` as the differentiator. I do not currently have a good suggestion other than to have `otRandomGet/Fill` as the prefix for non-cryptographic versions and `otRandomCryptoGet/Fill*` as the the prefix for cryptographic versions.

Thoughts?

Reply:
I was having same trouble - the other idea I had was `Crypto` and `Weak`

Reply:
I think we should put some keyword in software random generator functions so it makes a clear distinguishment between cryptographic and non-cryptographic functions. `Crypto` and `NonCrypto` is another option.

Reply:
`NonCrypto` works for me - I prefer that to `Software`.